ruby/sdk-lightweight.md

```diff
@@ -0,0 +1,462 @@
+
+
+## Types
+
+After many experiments we've decided to not proceed with Ruby typing at this point. This decision
+will be revisited when Ruby's typing system matures to support non-basic use-cases.
+
+
+## Defining an activity
+
+Activities are defined as classes ([a very common practice](https://guides.rubyonrails.org/active_job_basics.html)
+given that the language is object-oriented). Activities are expected to implement an `#execute`
+method that will be called to execute an activity.
+
+`sum_activity.rb`
+
+```ruby
+# The activity class name will be used as activity name unless overridden
+class SumActivity < Temporal::Activity
+  # Optionally override the name
+  activity_name 'sum-activity'
+
+  # return value of the #execute method will be used as an activity result
+  def execute(a, b)
+    a + b
+  end
+end
+```
+
+#### Alternative approaches
+
+- Using plain methods as activities. This is not feasible in Ruby since non-namespaced methods
+  are global without a clear way to reference them (would have to be `method(:my_activity)`)
+
+- It is possible to allow multiple activities per class, however we've decided against this because
+  there's an extra complexity to the invocation (extra argument representing the method) and it
+  complicates the use of class-level "decorators" (e.g. for specifying activity name or other
+  options)
+
+
+## Defining a workflow
+
+Similar to activities, workflows are also represented as classes.
+
+`sum_workflow.rb`
+
+```ruby
+require 'sum_activity'
+
+# Similar to activities the class name will be used as a workflow name unless overridden
+class SumWorkflow < Temporal::Workflow
+  # Optionally override the name
+  workflow_name 'sum-workflow'
+
+  def execute(a, b)
+    # - Positional arguments passed to activity (inline with other SDKs)
+    # - Keyword arguments are reserved for options
+    # - This call is blocking
+    workflow.execute_activity(SumActivity, a, b, start_to_close_timeout: 10)
+
+    # - Starting an activity by its name rather than a class
+    # - The result of this activity will be used as a workflow result (Ruby standard)
+    workflow.execute_activity('sum-activity', a, b, start_to_close_timeout: 10)
+  end
+end
+```
+
+#### Alternative approaches
+
+- For the same exact reasons it is not feasible to use non-namespaced methods for workflow
+
+- Similar to activities it should be possible to define multiple workflows per class, however this
+  will not be allowed to avoid a perception of a shared state between workflows and allow for
+  class-level "decorators"
+
+
+## Start a workflow
+
+```ruby
+require 'temporal/client'
+require 'sum_workflow'
+
+# Connections are eager and a lazy option will be introduced later
+connection = Temporal::Connection.connect('localhost:7233')
+client = Temporal::Client.new(connection, namespace: 'demo')
+
+# workflow is referenced by the class
+handle = client.start_workflow(SumWorkflow, 1, 2, id: 'id-1', task_queue: 'demo')
```

Comment:
task_queue: not sure if your intent here is to show the most common pattern.
With Nexus, temporal is moving in the direction of hiding the task queue from the caller wherever possible.
Even without Nexus though, it seems best if the workflow provides its own namespace and task queue whenever possible.  Can we do that?  I think only when you're calling indirectly via the name (without the class) would you strictly need to provide a task queue.  And I expect calling via class to be the 95% case.

Reply:
I don't think we can reasonably do this. Many users of the client don't even have the workflow definition present. All of our SDKs require task queue at workflow start time and I don't think Ruby should be unique here. Also, it is quite common for the same workflow to be on different task queues (e.g. task queue versioning).

As with other parts, users are encouraged to build on top of these primitives to wrap/centralize this stuff though. So maybe you have a static start method on your workflow class for Ruby users that knows the task queue to use.

Reply:
What I would recommend would be a way for a workflow/activity author to provide a default configuration that's usable by its in-language clients.  If clients can't use those defaults for whatever reason, they have to provide.  This could be two separate caller methods.  Many shops will be able to exclusively use the nice one.

Just because everybody can't have nice things doesn't mean most people can't.  And it'd be better for us to provide it than for everybody to have to roll their own--Maxim was saying as much at Replay.

Reply:
I think that may be a general discussion for all SDKs. I don't think just Ruby should deviate at this time.

Reply:
Agree with both points — it's a very useful property to have (and definitely made the current SDK much easier to use for many engineers) and also that this should come as a cross-SDK proposal.

@cretz are you aware of any ongoing efforts to propose this?

Reply:
I am not aware of any, but I'd of course support it. It'd need server side assistance and the upcoming Nexus framework expects to be able to support that as well (where the Nexus handler decides these options, not the caller).